### PR TITLE
`-Znext-solver`: eagerly normalize when adding goals

### LINKED ISSUE #125269: endless loop by running `cargo test` with toolchain 2024-05-14
I tried this code:

(sorry, i couldn't extract a minimal example and reproduce it)

```bash
git clone https://gitlab.com/veloren/veloren
cd veloren
git checkout e48fc4c2b235ceab3c1a551292c45c46d47e7130
cd common
cargo test
```

I expected to see this happen: ElimCase doctest in line 86 to finish after some time

Instead, this happened: the tests takes endless (i waited 2 hours in CI)

### Meta
`rustc --version --verbose`:
```
rustc 1.80.0-nightly (ab14f944a 2024-05-13)
binary: rustc
commit-hash: ab14f944afe4234db378ced3801e637eae6c0f30
commit-date: 2024-05-13
host: x86_64-unknown-linux-gnu
release: 1.80.0-nightly
LLVM version: 18.1.4
```

`cargo test`:
```
   Doc-tests veloren_common

running 14 tests
test common/src/terrain/block.rs - terrain::block::Block (line 120) ... ignored
...
test common/src/typed.rs - typed::ElimCase (line 107) ... ok
...
test common/src/typed.rs - typed::ElimCase (line 86) has been running for over 60 seconds
```

#### code in the doc-test in question:
```
/// ```
/// veloren_common::make_case_elim!(
///     my_type_module,
///     #[repr(u32)]
///     #[derive(Clone,Copy)]
///     pub enum MyType {
///         Constr1 = 0,
///         #[typed(pure)] Constr2(arg : u8) = 1,
///         /* ..., */
///     }
/// );
/// ```
```
### Workaround
As workaround we added a main function to the doctest, as the test doesnt execute code so far (only defines stuff) it then runs immediately again
https://gitlab.com/veloren/veloren/-/commit/3cc4b947ebd049dcdf82ca4a66a815d32d4d43f8

### PR DESCRIPTION
**Pull request by lcnr**
fixes #125269. I am not totally with this fix and going to keep this open until we have a more general discussion about how to handle hangs caused by lazy norm in the new solver.

### CODE REVIEW

File: compiler/rustc_trait_selection/src/solve/eval_ctxt/mod.rs [hunk 1]
```diff
@@ -1101,3 +1114,37 @@ impl<'tcx> EvalCtxt<'_, InferCtxt<'tcx>> {
...
+                );
+                self.ecx.add_goal(
+                    GoalSource::Misc,
+                    Goal::new(self.interner(), self.param_env, normalizes_to),
+                );
+                infer_ty
+            }
+            _ => ty.super_fold_with(self),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/solve/eval_ctxt/mod.rs:1145]: We should try to minimize the number of places where we treat consts and types differently, even if we never expect this code to be hit, if anything to make sure future work on unevaluated consts doesn't have to remember to do this later.

**Comment by compiler-errors** [comment]
We discussed this, I'm happy with landing this for now I guess.

We should track this optimization somewhere (ideally with a label in the new solver repo) so we can revisit it.

@bors r+

**Comment by lcnr** [comment]
opened https://github.com/rust-lang/trait-system-refactor-initiative/issues/113

**Comment by lqd** [comment]
@lcnr how about the hang we found in this PR when using `-Znext-solver=coherence`?

**Pull request merged**